Add a HttpPoller that remembers failures

JacekLach opened the pull request:
At times it is reasonable to wait for a long (~minutes) time for resources
to become available - think large docker-composes on limited hardware.

This requires configuring the default poller with a fairly long timeout. Unfortunately
if you do that, and one of the resources is misconfigured and will never come up, you
will wait the entire long timeout for each test in your suite before aborting the entire
build. This can easily take you from a build taking ~10 minutes to over an hour.

The FailureCachingHttpPollingResource allows you to tolerate a long wait as long as the
resource becomes available eventually; but once you do time out, any subsequent tests will
fail eagerly instead of waiting the full timeout again.

JacekLach committed:
Add a HttpPoller that remembers failures

At times it is reasonable to wait for a long (~minutes) time for resources
to become available - think large docker-composes on limited hardware.

This requires configuring the default poller with a fairly long timeout. Unfortunately
if you do that, and one of the resources is misconfigured and will never come up, you
will wait the entire long timeout for each test in your suite before aborting the entire
build. This can easily take you from a build taking ~10 minutes to over an hour.

The FailureCachingHttpPollingResource allows you to tolerate a long wait as long as the
resource becomes available eventually; but once you do time out, any subsequent tests will
fail eagerly instead of waiting the full timeout again.

JacekLach commented:
CC @bavardage @uschi2000 - I can't assign reviewers.

uschi2000 commented on junit-resource-poller/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
why do we need this?

bavardage replied:
I guess to allow mocking of the final class `HttpPollingResource`

JacekLach replied:
Yep, `HttpPollingResource` is unfortunately final.

Alternative is to relax `FailureCachingHttpPollingResource` to take in any `ExternalResource` as a delegate, but I'm wary of that; we're pretty tightly coupled to what `HttpPollingResource` does (noop `after()`,  skippable `before()`)

bavardage commented on junit-resource-poller/src/main/java/com/palantir/junit/FailureCachingHttpPollingResource.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public final class FailureCachingHttpPollingResource extends ExternalResource {
+    private final HttpPollingResource poller;
+    private AtomicReference<Throwable> maybeError = new AtomicReference<>();
```
can this be final?

JacekLach replied:
Done

JacekLach committed:
Do not lazySet throwable reference

We're never in a context where the performance impact of a storeload barrier is
significant. Using AtomicReference#set instead of #lazySet makes reasoning about
behaviour of concurrent tests much easier.

JacekLach committed:
Fix checkstyle
